File: backend/app/schemas/user_management.py
"""User management schemas (master only)."""
from datetime import datetime
from typing import Literal, Optional
from pydantic import BaseModel, ConfigDict, field_validator

from app.schemas.user import UserResponse


class UserWithActivityResponse(UserResponse):
    teacher_name: Optional[str] = None
    last_active: Optional[datetime] = None
    total_sessions: int = 0
    accuracy_pct: Optional[float] = None


class UserListResponse(BaseModel):
    users: list[UserWithActivityResponse]
    total: int
    page: int
    page_size: int


class CreateUserRequest(BaseModel):
    name: str
    username: str
    password: str
    role: Literal["student", "teacher", "master"]
    email: Optional[str] = None
    phone_number: Optional[str] = None
    school_name: Optional[str] = None
    grade: Optional[str] = None
    teacher_id: Optional[str] = None

    @field_validator("name")
    @classmethod
    def name_not_empty(cls, v: str) -> str:
        if not v or not v.strip():
            raise ValueError("이름을 입력해주세요")
        return v.strip()

    @field_validator("password")
    @classmethod
    def password_min_length(cls, v: str) -> str:
        if len(v) < 4:
            raise ValueError("비밀번호는 4자 이상이어야 합니다")
        return v


class UpdateUserRequest(BaseModel):
    name: Optional[str] = None
    email: Optional[str] = None
    password: Optional[str] = None
    phone_number: Optional[str] = None
    school_name: Optional[str] = None
    grade: Optional[str] = None
    role: Optional[str] = None
    teacher_id: Optional[str] = None


class ResetPasswordResponse(BaseModel):
    temporary_password: str


class UserDetailSessionItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    session_id: str
    session_type: str
    started_at: datetime
    completed_at: Optional[datetime] = None
    total_questions: int
    correct_count: int
    accuracy_pct: Optional[float] = None
    duration_seconds: Optional[int] = None


class UserDetailResponse(BaseModel):
    user: UserWithActivityResponse
    sessions: list[UserDetailSessionItem]

File: frontend/src/services/userManagement.ts
/**
 * User management API service (master only).
 */
import api from './api';

export interface UserWithActivity {
  id: string;
  email: string | null;
  username: string | null;
  name: string;
  role: string;
  teacher_id: string | null;
  teacher_name: string | null;
  school_name: string | null;
  grade: string | null;
  phone_number: string | null;
  created_at: string;
  updated_at: string;
  last_active: string | null;
  total_sessions: number;
  accuracy_pct: number | null;
}

export interface UserListResponse {
  users: UserWithActivity[];
  total: number;
  page: number;
  page_size: number;
}

export interface UserListParams {
  page?: number;
  page_size?: number;
  search?: string;
  role?: string;
  sort_by?: string;
  sort_dir?: 'asc' | 'desc';
}

export interface CreateUserRequest {
  name: string;
  username: string;
  password: string;
  role: 'student' | 'teacher' | 'master';
  email?: string;
  phone_number?: string;
  school_name?: string;
  grade?: string;
  teacher_id?: string;
}

export interface UpdateUserRequest {
  name?: string;
  email?: string;
  password?: string;
  phone_number?: string;
  school_name?: string;
  grade?: string;
  role?: string;
  teacher_id?: string;
}

export interface UserDetailSession {
  session_id: string;
  session_type: 'learning' | 'grammar' | 'legacy';
  started_at: string;
  completed_at: string | null;
  total_questions: number;
  correct_count: number;
  accuracy_pct: number | null;
  duration_seconds: number | null;
}

export interface UserDetailResponse {
  user: UserWithActivity;
  sessions: UserDetailSession[];
}

export interface ResetPasswordResponse {
  temporary_password: string;
}

export const userManagementService = {
  async listUsers(params?: UserListParams): Promise<UserListResponse> {
    const response = await api.get<UserListResponse>('/api/v1/user-management/', { params });
    return response.data;
  },

  async getUserDetail(id: string): Promise<UserDetailResponse> {
    const response = await api.get<UserDetailResponse>(`/api/v1/user-management/${id}`);
    return response.data;
  },

  async createUser(data: CreateUserRequest): Promise<UserWithActivity> {
    const response = await api.post<UserWithActivity>('/api/v1/user-management/', data);
    return response.data;
  },

  async updateUser(id: string, data: UpdateUserRequest): Promise<UserWithActivity> {
    const response = await api.patch<UserWithActivity>(`/api/v1/user-management/${id}`, data);
    return response.data;
  },

  async deleteUser(id: string): Promise<void> {
    await api.delete(`/api/v1/user-management/${id}`);
  },

  async resetPassword(id: string): Promise<ResetPasswordResponse> {
    const response = await api.post<ResetPasswordResponse>(`/api/v1/user-management/${id}/reset-password`);
    return response.data;
  },
};

export default userManagementService;
